test-data/unit/check-attr.test

```diff
@@ -0,0 +1,444 @@
+    a: int
+reveal_type(A)  # E: Revealed type is 'def (a: builtins.int) -> __main__.A'
+reveal_type(A.__eq__)  # E: Revealed type is 'def (builtins.object, builtins.object) -> builtins.bool'
+A(1) < A(2)  # E: Unsupported left operand type for < ("A")
+[builtins fixtures/attr.pyi]
+
+[case testAttrsInheritance]
+import attr
+@attr.s
+class A:
+    a: int = attr.ib()
+@attr.s
+class B:
+    b: str = attr.ib()
+@attr.s
+class C(A, B):
+    c: bool = attr.ib()
+reveal_type(C)  # E: Revealed type is 'def (a: builtins.int, b: builtins.str, c: builtins.bool) -> __main__.C'
+[builtins fixtures/bool.pyi]
+
+[case testAttrsNestedInClasses]
+import attr
+@attr.s
+class C:
+    y = attr.ib()
+    @attr.s
+    class D:
+        x: int = attr.ib()
+reveal_type(C)  # E: Revealed type is 'def (y: Any) -> __main__.C'
+reveal_type(C.D)  # E: Revealed type is 'def (x: builtins.int) -> __main__.C.D'
+[builtins fixtures/bool.pyi]
+
+[case testAttrsInheritanceOverride]
+import attr
+
+@attr.s
+class A:
+    a: int = attr.ib()
+    x: int = attr.ib()
+
+@attr.s
+class B(A):
+    b: str = attr.ib()
+    x: int = attr.ib(default=22)
+
+@attr.s
+class C(B):
+    c: bool = attr.ib()  # No error here because the x below overwrites the x above.
+    x: int = attr.ib()
+
+reveal_type(A)  # E: Revealed type is 'def (a: builtins.int, x: builtins.int) -> __main__.A'
+reveal_type(B)  # E: Revealed type is 'def (a: builtins.int, b: builtins.str, x: builtins.int =) -> __main__.B'
+reveal_type(C)  # E: Revealed type is 'def (a: builtins.int, b: builtins.str, c: builtins.bool, x: builtins.int) -> __main__.C'
+[builtins fixtures/bool.pyi]
+
+[case testAttrsTypeEquals]
+import attr
+
+@attr.s
+class A:
+    a = attr.ib(type=int)
+    b = attr.ib(18, type=int)
+reveal_type(A) # E: Revealed type is 'def (a: builtins.int, b: builtins.int =) -> __main__.A'
+[builtins fixtures/bool.pyi]
+
+[case testAttrsFrozen]
+import attr
+
+@attr.s(frozen=True)
+class A:
+    a = attr.ib()
+
+a = A(5)
+a.a = 16  # E: Property "a" defined in "A" is read-only
+[builtins fixtures/bool.pyi]
+
+[case testAttrsDataClass]
+import attr
+from typing import List, ClassVar
+@attr.dataclass
+class A:
+    a: int
+    _b: List[str]
+    c: str = '18'
+    _d: int = attr.ib(validator=None, default=18)
+    E = 7
+    F: ClassVar[int] = 22
+reveal_type(A)  # E: Revealed type is 'def (a: builtins.int, b: builtins.list[builtins.str], c: builtins.str =, d: builtins.int =) -> __main__.A'
+A(1, ['2'])
+[builtins fixtures/list.pyi]
+
+[case testAttrsTypeAlias]
+from typing import List
+import attr
+Alias = List[int]
+@attr.s(auto_attribs=True)
+class A:
+    Alias2 = List[str]
+    x: Alias
+    y: Alias2 = attr.ib()
+reveal_type(A) # E: Revealed type is 'def (x: builtins.list[builtins.int], y: builtins.list[builtins.str]) -> __main__.A'
+[builtins fixtures/list.pyi]
+
+[case testAttrsTypeVariable]
+from typing import TypeVar, Generic, List
+import attr
+T = TypeVar('T')
+@attr.s(auto_attribs=True)
+class A(Generic[T]):
```

Comment:
Also please add a test where generic class has a method, and it contains access to an attribute on ``self``.